liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,125 @@
+module Language.Haskell.Liquid.Constraint.RewriteCase 
+    (getCaseRewrites) 
+    where
+
+import           Language.Fixpoint.Types
+import qualified Language.Fixpoint.Misc as M
+import           Language.Haskell.Liquid.Constraint.Types
+import           Language.Haskell.Liquid.Types.Types
+import           Language.Haskell.Liquid.Types.RType
+
+import           Data.Maybe
+import           Control.Monad
+
+import qualified Data.HashMap.Strict as M
+import qualified Data.HashSet        as S
+
+getCaseRewrites :: CGEnv -> SpecType -> LocalRewrites
+getCaseRewrites γ spec =
+    let Reft (_, refinement) = ur_reft $ rt_reft spec
+        -- Names of constustors both type constructors and data constructors
+        ctors   = toSet $ seBinds  $ constEnv γ
+        -- All the global names (top level functions, types, etc...)
+        globals = toSet $ reGlobal $ renv     γ
+    in unloop
+       $ concat
+       $ mapMaybe (uncurry $ unify ctors globals)
+       $ groupUnifiableEqualities 
+       $ getEqualities refinement
+    where toSet = S.fromList . M.keys
+
+unify :: S.HashSet Symbol -> S.HashSet Symbol -> Expr -> Expr -> Maybe [(Symbol, Expr)]
+unify ctors globals = go
+    where
+        go e1 e2 | e1 == e2 = Just []
+        -- NOTE: We don't need to check for ECst because the expressions arent
+        -- elaborated
+        go (EVar s1) e2 | isLocal s1 = Just [(s1, e2)]
+        go e1 (EVar s2) | isLocal s2 = Just [(s2, e1)]
+        -- TODO: Tecnically we could also unify under lambdas but you have to be
+        -- carefull about alpha equivalence idk if the effort is worth it.
+        go e1 e2 = do
+            -- Performing the unification under constructor is safe because 
+            -- C a₁ ... aₙ = C b₁ ... bₙ ⟺ ∀ n . a₁ = bₙ
+            EVar name1 : args1 <- pure $ flatten e1
+            EVar name2 : args2 <- pure $ flatten e2
+            guard $ name1 == name2
+            guard $ isCtor name1
+            guard $ length args1 == length args2
+            concat <$> zipWithM go args1 args2
```

Comment:
The code would be somewhat simplified if the return type of `unify` is changed from `Maybe [(Symbol, Expr)]` to just a list `[(Symbol, Expr)]` which should be empty when no unifications are found. In that case, the final statement here turns to
```
            concat $ zipWith go args1 args2
```

Reply:
The unification Just [] is different from Nothing, Nothing means the unification failed so we bail completely, instead Just [] means the unification succeeded but we didn't discovered any equality

Reply:
The only use of `unify` doesn't distinguish between those two cases. But I'm guessing you mean to leave it as it is in case it makes a difference in the future. That would be fine by me.

Reply:
zipWithM should collapse to Nothing if one of the unification is returning Nothing

Reply:
Ok, so `C (f a) (g x)` does not unify with `C (h a) y`. I would think we could use `y=g x` despite the failed unification of the first argument.

Reply:
Uhm technically yes I don't know why I wrote it like that